nevergrad/functions/multiobjective.py

```diff
@@ -0,0 +1,387 @@
+#    TU Dortmund University
+#
+#    In personal communication, the original authors authorized DEAP team
+#    to use this file under the Lesser General Public License.
+#
+#    You can find the original library here :
+#    http://ls11-www.cs.uni-dortmund.de/_media/rudolph/hypervolume/hv_python.zip
+#
+#    DEAP is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Lesser General Public License as
+#    published by the Free Software Foundation, either version 3 of
+#    the License, or (at your option) any later version.
+#
+#    DEAP is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+#    GNU Lesser General Public License for more details.
+#
+#    You should have received a copy of the GNU Lesser General Public
+#    License along with DEAP. If not, see <http://www.gnu.org/licenses/>.
+#
+#    Modified by O. Teytaud as needed for Nevergrad.
+
+from ..instrumentation import InstrumentedFunction
+from math import log, floor
+import random
+from typing import (Tuple, Any, Callable, List, Optional, Dict, ValuesView, Iterator,
+                    TypeVar, Generic, Deque, Iterable)
+import warnings
+import numpy
+#from ..optimization import base
+from ..optimization.base import Candidate
+from ..common.typetools import ArrayLike
+from ..instrumentation import InstrumentedFunction
+
+def hypervolume(pointset, ref):
+    """Compute the absolute hypervolume of a *pointset* according to the
+    reference point *ref*.
+    """
+    hv = _HyperVolume(ref)
+    return hv.compute(pointset)
+
+
+class MultiobjectiveFunction(InstrumentedFunction):
+    """Given several functions, and threshold on their values (above which solutions are pointless),
+    this function returns a single-objective function, correctly instrumented, the minimization of which
+    yields a solution to the original multiobjective problem.
+    
+    multi_objective_function: objective functions, to be minimized, of the original multiobjective problem.
+    upper_bounds: upper_bounds[i] is a threshold above which x is pointless if functions[i](x) > upper_bounds[i].
+    
+    Returns an objective function to be minimized (it is a single objective function).
+    Warning: this function is not stationary.
+    The minimum value obtained for this objective function is -h,
+    where h is the hypervolume of the Pareto front obtained, given upper_bounds as a reference point.
+    """
+
+
+    def __init__(self, multiobjective_function: Callable[..., Tuple[float, ...]], upper_bounds: Tuple[float, ...]) -> None:
+        self.multi_objective_function = multiobjective_function
+        self.upper_bounds = numpy.array(upper_bounds)
+        self.pointset: Dict[Candidate, ArrayLike] = {}
+        self.best_hypervolume = -float("Inf")
+
+    def compute_aggregate_loss(self, x: Candidate, v: Tuple[float]) -> None:
```

Comment:
what is v?? losses? the name is not explicit enough
the signature should be `losses: Tuple[float,...], *args: Any, **kwargs: Any`

Reply:
v is the objective function values, i.e. the losses